src/gc-heap-snapshot.cpp

```diff
@@ -0,0 +1,495 @@
+        snapshot->names.find_or_create_string_id("(stack frame)"), // name
+        (size_t)frame, // id
+        1, // size
+        0, // size_t trace_node_id (unused)
+        0, // int detachedness;  // 0 - unknown,  1 - attached;  2 - detached
+        vector<Edge>() // outgoing edges
+    });
+
+    return val.first->second;
+}
+
+void _gc_heap_snapshot_record_frame_to_object_edge(void *from, jl_value_t *to) JL_NOTSAFEPOINT {
+    auto from_node_idx = _record_stack_frame_node(g_snapshot, (jl_gcframe_t*)from);
+    auto to_idx = record_node_to_gc_snapshot(to);
+    Node &from_node = g_snapshot->nodes[from_node_idx];
+
+    auto name_idx = g_snapshot->names.find_or_create_string_id("local var");
+    _record_gc_just_edge("internal", from_node, to_idx, name_idx);
+}
+
+void _gc_heap_snapshot_record_task_to_frame_edge(jl_task_t *from, void *to) JL_NOTSAFEPOINT {
+    auto from_node_idx = record_node_to_gc_snapshot((jl_value_t*)from);
+    auto to_node_idx = _record_stack_frame_node(g_snapshot, to);
+    Node &from_node = g_snapshot->nodes[from_node_idx];
+
+    auto name_idx = g_snapshot->names.find_or_create_string_id("stack");
+    _record_gc_just_edge("internal", from_node, to_node_idx, name_idx);
+}
+
+void _gc_heap_snapshot_record_frame_to_frame_edge(jl_gcframe_t *from, jl_gcframe_t *to) JL_NOTSAFEPOINT {
+    auto from_node_idx = _record_stack_frame_node(g_snapshot, from);
+    auto to_node_idx = _record_stack_frame_node(g_snapshot, to);
+    Node &from_node = g_snapshot->nodes[from_node_idx];
+
+    auto name_idx = g_snapshot->names.find_or_create_string_id("next frame");
+    _record_gc_just_edge("internal", from_node, to_node_idx, name_idx);
+}
+
+void _gc_heap_snapshot_record_array_edge(jl_value_t *from, jl_value_t *to, size_t index) JL_NOTSAFEPOINT {
+    _record_gc_edge("array", "element", from, to, index);
+}
+
+void _gc_heap_snapshot_record_module_edge(jl_module_t *from, jl_value_t *to, char *name) JL_NOTSAFEPOINT {
+    _record_gc_edge("object", "property", (jl_value_t *)from, to,
+                    g_snapshot->names.find_or_create_string_id(name));
+}
+
+void _gc_heap_snapshot_record_object_edge(jl_value_t *from, jl_value_t *to, void* slot) JL_NOTSAFEPOINT {
+    string path = _fieldpath_for_slot(from, slot);
+    _record_gc_edge("object", "property", from, to,
+                    g_snapshot->names.find_or_create_string_id(path));
+}
+
+void _gc_heap_snapshot_record_module_to_binding(jl_module_t* module, jl_binding_t* binding) JL_NOTSAFEPOINT {
+    
+
+    auto from_node_idx = record_node_to_gc_snapshot((jl_value_t*)module);
+    auto to_node_idx = record_pointer_to_gc_snapshot(binding, sizeof(jl_binding_t), jl_symbol_name(binding->name));
+    auto value_idx = (binding->value) ? record_node_to_gc_snapshot(binding->value) : 0;
+    auto ty_idx = (binding->ty) ? record_node_to_gc_snapshot(binding->ty) : 0;
+    auto globalref_idx = (binding->globalref) ? record_node_to_gc_snapshot(binding->globalref) : 0;
```

Comment:
These are all `_Atomic` variables, so loading them more than once will fail our atomics usage static checker. Fix that by introducing a local copy of the value, like so:
```suggestion
    jl_value_t *value = jl_atomic_load_relaxed(&binding->value);
    auto value_idx = value ? record_node_to_gc_snapshot(value) : 0;
```